File: frontend/src/lib/valuation/par.ts
// Port of src/fantasy_value/valuation/par.py.
//
// Use `??`, never `||`. A replacement level of 0 is legitimate (a position
// where the boundary player scores exactly zero). `levels.get(pos) || null`
// would silently convert that 0 to null and erase every PAR at that
// position.

import type { ScoredPlayer } from './models'

export function computePar(player: ScoredPlayer, replacementLevels: ReadonlyMap<string, number | null>): number | null {
  const level = replacementLevels.get(player.position) ?? null
  return level === null ? null : player.points - level
}

export function computeParById(
  pool: Iterable<ScoredPlayer>,
  replacementLevels: ReadonlyMap<string, number | null>,
): Map<string, number | null> {
  const result = new Map<string, number | null>()
  for (const p of pool) {
    result.set(p.player_id, computePar(p, replacementLevels))
  }
  return result
}

File: src/fantasy_value/valuation/par.py
from __future__ import annotations

from typing import Iterable, Mapping

from fantasy_value.valuation.models import ScoredPlayer


def compute_par(player: ScoredPlayer, replacement_levels: Mapping[str, float | None]) -> float | None:
    """Section 16: PAR = points - replacement level. Propagates None (an
    undefined replacement level) rather than crashing or defaulting to 0."""
    level = replacement_levels.get(player.position)
    return None if level is None else player.points - level


def compute_par_by_id(
    pool: Iterable[ScoredPlayer], replacement_levels: Mapping[str, float | None]
) -> dict[str, float | None]:
    return {p.player_id: compute_par(p, replacement_levels) for p in pool}
